### PR TITLE
feat(slider)!: add a groupSeparator prop to enable toggling display of separator in label

### PR DESCRIPTION
**Pull request by Elijbet**
**Related Issue:** #4976

## Summary

Adds a `groupSeparator` prop to `slider` which displays numbers as separated and localized when true (e.g. 1,970 for 'en') and does not when false/not present (e.g. 1970). `lang` and `numberingSystem` props are added to support `localizeNumberString` util used for handling the localization.

Before this change, `slider` was automatically adding the language's corresponding number separator to the value label.

Adds a number locale support test to check for tick min/max and thumb min/max localized display values based on `groupSeparator` prop.

### CODE REVIEW

File: src/components/slider/slider.tsx [hunk 1]
```diff
@@ -150,6 +150,16 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
    */
   @Prop() scale: Scale = "m";
 
+  /**
+   * When true, displays numbers as comma separated (e.g. 1,970) and when false/not present does not (e.g. 1970)
+   */
+  @Prop({ reflect: true, mutable: true }) groupSeparator = false;
+
+  @Watch("groupSeparator")
```
Comment (driskull) [inline src/components/slider/slider.tsx:158]: @Elijbet I don't think this watcher is necessary as it's just setting the value to what it has already been set as.

File: src/components/slider/slider.tsx [hunk 2]
```diff
@@ -574,13 +584,13 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
...
-          {this.minValue && this.minValue.toLocaleString()}
+          {this.minValue && this.determineGroupSeparator(this.minValue)}
         </span>
         <span aria-hidden="true" class="handle__label handle__label--minValue static">
-          {this.minValue && this.minValue.toLocaleString()}
+          {this.minValue && this.determineGroupSeparator(this.minValue)}
         </span>
         <span aria-hidden="true" class="handle__label handle__label--minValue transformed">
-          {this.minValue && this.minValue.toLocaleString()}
+          {this.minValue && this.determineGroupSeparator(this.minValue)}
```
Comment (driskull) [inline src/components/slider/slider.tsx:593]: Instead of all the `this.minValue && ` maybe `determineGroupSeparater` could handle when the value is not set?

File: src/components/slider/slider.tsx [hunk 3]
```diff
@@ -1396,4 +1406,13 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
...
     return intersects(maxLabelBounds, handleBounds);
   }
+
+  /**
+   * Returns a string representing the label value that is parsed for comma separator based on whether groupSeparator prop is set to true/false.
+   *
+   * @param input
+   */
+  private determineGroupSeparator = (input: number): string => {
+    return this.groupSeparator ? input.toLocaleString() : input.toLocaleString().replace(",", "");
```
Comment (driskull) [inline src/components/slider/slider.tsx:1416]: question: Does `input.toLocaleString()` always have commas or could there be other number separators for different languages?

File: src/components/slider/slider.tsx [hunk 4]
```diff
@@ -574,13 +584,13 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
         <div class="handle-extension" />
         <div class="handle" />
         <span aria-hidden="true" class="handle__label handle__label--minValue">
-          {this.minValue && this.minValue.toLocaleString()}
+          {this.minValue && this.determineGroupSeparator(this.minValue)}
```
Comment (driskull) [inline src/components/slider/slider.tsx:587]: The `minValue` of `0` won't call `determineGroupSeparator` Is this ok?

**Review by driskull — COMMENTED** [review]
Code looks good! Have some comments/questions regarding a few things.

File: src/components/slider/slider.e2e.ts [hunk 5]
```diff
@@ -673,4 +673,49 @@ describe("calcite-slider", () => {
   it("is form-associated", () => formAssociated("calcite-slider", { testValue: 5 }));
 
   it("is form-associated with range", () => formAssociated("calcite-slider", { testValue: [5, 10] }));
+
+  it("groupSeparator: displays numbers as comma separated when prop is true and does not when false", async () => {
+    const page = await newE2EPage({
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:678]: For consistency, can you create an empty page and use `setContent` afterwards? No need for `waitForChanges` after doing this.

File: src/components/slider/slider.e2e.ts [hunk 6]
```diff
@@ -673,4 +673,49 @@ describe("calcite-slider", () => {
...
+      const labelVal = (await element.shadowRoot.querySelector("span.handle__label--value")) as HTMLElement;
+
+      const tickMin = (await element.shadowRoot.querySelector("span.tick__label--min")) as HTMLElement;
+      const tickMax = (await element.shadowRoot.querySelector("span.tick__label--max")) as HTMLElement;
+
+      return [labelMinVal.innerText, labelVal.innerText, tickMin.innerText, tickMax.innerText];
+    };
+    await page.exposeFunction("displayedValuesArray", displayedValuesArray);
+
+    const noSeparator = await page.$eval("calcite-slider", async (): Promise<string[]> => {
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:707]: I think you can get rid of the wrapping anonymous functions and pass `displayedValuesArray` directly.

File: src/components/slider/slider.e2e.ts [hunk 7]
```diff
@@ -673,4 +673,49 @@ describe("calcite-slider", () => {
...
+                label-handles 
+                label-ticks 
+                style="width:${sliderWidthFor1To1PixelValueTrack}">
+            </calcite-slider>`
+    });
+    await page.waitForChanges();
+    const element = await page.find("calcite-slider");
+
+    const displayedValuesArray = async (): Promise<string[]> => {
+      await page.waitForChanges();
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:695]: I think you can get rid of this since you already wait for changes before calling this method.

File: src/components/slider/slider.e2e.ts [hunk 8]
```diff
@@ -673,4 +673,49 @@ describe("calcite-slider", () => {
...
+    });
+
+    element.setProperty("groupSeparator", true);
+    await page.waitForChanges();
+
+    const withSeparator = await page.$eval("calcite-slider", async (): Promise<string[]> => {
+      return await displayedValuesArray();
+    });
+
+    expect([...noSeparator]).toEqual(["2500", "5000", "1000", "10000"]);
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:718]: You can pass the arrays directly and not need to spread values.

File: src/components/slider/slider.tsx [hunk 9]
```diff
@@ -257,13 +262,13 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
         tabIndex={0}
       >
         <span aria-hidden="true" class="handle__label handle__label--value">
-          {value ? value.toLocaleString() : value}
+          {value ? this.determineGroupSeparator(value) : value}
```
Comment (jcfranco) [inline src/components/slider/slider.tsx:265]: Optional: can you compute the value once, store in a local variable and reuse?

File: src/components/slider/slider.tsx (hunk 2 above)
Comment (jcfranco) [reply src/components/slider/slider.tsx:588]: I agree with the suggestion. Looks like it was guarding against it to avoid calling a number on `null`/`undefined`.

File: src/components/slider/slider.tsx (hunk 3 above)
Comment (jcfranco) [reply src/components/slider/slider.tsx:1411]: Can you use [`localizeNumberString` from `src/utils/locale.ts`](https://github.com/Esri/calcite-components/blob/master/src/utils/locale.ts#L106) instead? It should only be used if `groupSeparator` is true and return the value as is otherwise. +@eriklharper, @benelan for additional hints

File: src/components/slider/slider.tsx [hunk 10]
```diff
@@ -1396,4 +1401,13 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
     const handleBounds = handle.getBoundingClientRect();
     return intersects(maxLabelBounds, handleBounds);
   }
+
+  /**
+   * Returns a string representing the label value that is parsed for comma separator based on whether groupSeparator prop is set to true/false.
+   *
+   * @param input
+   */
+  private determineGroupSeparator = (input: number): string => {
```
Comment (jcfranco) [inline src/components/slider/slider.tsx:1410]: Can you rename this method to better express what it does? The current name doesn't tell me that it returns a formatted value.

Also, can you rename `input` to `value` or something of that nature?

File: src/components/slider/slider.e2e.ts [hunk 11]
```diff
@@ -673,4 +673,49 @@ describe("calcite-slider", () => {
...
+    expect([...withSeparator]).toEqual(["2,500", "5,000", "1,000", "10,000"]);
+  });
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:720]: Great test! Can you update it to cover additional locales? Something like https://github.com/Esri/calcite-components/blob/adcbb55b3ab3ab9fca0921b4bbbea1e4bf92907b/src/components/input/input.e2e.ts#L1057 would be good.

**Review by jcfranco — CHANGES_REQUESTED** [review]

File: src/components/slider/slider.e2e.ts (hunk 5 above)
Comment (Elijbet) [reply src/components/slider/slider.e2e.ts:678]: Oops, sorry.

File: src/components/slider/slider.tsx (hunk 3 above)
Comment (Elijbet) [reply src/components/slider/slider.tsx:1411]: > commas

 Yeah, good point. A period is used as the thousands separator in many non-English speaking countries.

File: src/components/slider/slider.tsx [hunk 12]
```diff
@@ -150,6 +151,21 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
    */
   @Prop() scale: Scale = "m";
 
+  /**
+   * When true, number values are displayed with the locale's group separator.
+   */
+  @Prop({ reflect: true }) groupSeparator = false;
+
+  /** Specifies the BCP 47 language tag for the desired language and country format. */
+  @Prop() locale: string = document.documentElement.lang || "en";
```
Comment (driskull) [inline src/components/slider/slider.tsx:160]: These might be related to this:

https://github.com/Esri/calcite-components/issues/4576
https://github.com/Esri/calcite-components/issues/4213

Should we start prioritizing these? Ideally, we should just use the global `lang` attribute.

**Review by driskull — COMMENTED** [review]
Looks good @Elijbet!!!

My only concern is if we should not introduce the `locale` property here since we are planning on deprecating it elsewhere.

File: src/components/slider/slider.tsx (hunk 12 above)
Comment (Elijbet) [reply src/components/slider/slider.tsx:160]: Ok, I'll make the change to `lang` instead.

Comment (jcfranco) [reply src/components/slider/slider.tsx:160]: These are scheduled for this month according to the roadmap. 😄

Comment (jcfranco) [reply src/components/slider/slider.tsx:160]: In this PR's context, let's move forward w/ `lang`. I forgot `pagination` already was using our global util, so let's use that moving forward and we'll align other components this month.

File: src/components/slider/slider.tsx [hunk 13]
```diff
@@ -150,6 +151,21 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
    */
   @Prop() scale: Scale = "m";
 
+  /**
+   * When true, number values are displayed with the locale's group separator.
+   */
+  @Prop({ reflect: true }) groupSeparator = false;
```
Comment (jcfranco) [inline src/components/slider/slider.tsx:157]: Can you move this prop so it's sorted alphabetically?

**Review by jcfranco — APPROVED** [review]
Aside from a few comments, this LGTM! 🎉🌮🦖🦃

File: src/components/slider/slider.e2e.ts [hunk 14]
```diff
@@ -673,4 +673,112 @@ describe("calcite-slider", () => {
...
+      element.setProperty("locale", "hi");
+      await page.waitForChanges();
+
+      withSeparator = await page.$eval("calcite-slider", async (): Promise<string[]> => {
+        return await getDisplayedValuesArray();
+      });
+      expect(withSeparator).toEqual(["2,500", "5,00,000.5", "1,000", "10,00,000.5"]);
+    });
+
+    it("Spanish", async () => {
```
[the remark this answers is not in the document]
Comment (Elijbet) [reply src/components/slider/slider.e2e.ts:774]: It all gotta Spanish!

File: src/components/slider/slider.e2e.ts [hunk 15]
```diff
@@ -673,4 +673,112 @@ describe("calcite-slider", () => {
...
+      await page.waitForChanges();
+
+      noSeparator = await page.$eval("calcite-slider", async (): Promise<string[]> => {
+        return await getDisplayedValuesArray();
+      });
+      expect(await element.getProperty("groupSeparator")).toBe(false);
+      expect(noSeparator).toEqual(expectedNotSeparatedValueArray);
+    });
+
+    it("English", async () => {
```
[the remark this answers is not in the document]
Comment (Elijbet) [reply src/components/slider/slider.e2e.ts:734]: **### 🧽🧼🍽✨**

File: src/components/slider/resources.ts [hunk 16]
```diff
@@ -0,0 +1,6 @@
+export const CSS = {
+  labelMinValue: "span.handle__label--minValue",
```
Comment (driskull) [inline src/components/slider/resources.ts:2]: how come these have `span` inside them? It doesn't seem like we should need to have the element type in the class

File: src/components/slider/slider.tsx [hunk 17]
```diff
@@ -1396,4 +1423,18 @@ export class Slider implements LabelableComponent, FormComponent, InteractiveCom
...
     return intersects(maxLabelBounds, handleBounds);
   }
+
+  /**
+   * Returns a string representing the localized label value based on groupSeparator prop being on or off.
+   *
+   * @param value
+   */
+  private determineGroupSeparator = (value): string => {
+    const lang = this.globalAttributes["lang"] || document.documentElement.lang || "en";
```
Comment (driskull) [inline src/components/slider/slider.tsx:1433]: I wonder instead of `document.documentElement.lang` we should do `this.el.closes(["lang"]).lang`? What do you think @jcfranco @Elijbet ? Right now, it doesn't honor parent elements lang.

I'm ok with enhancing this in globalAttributes helper if we can do that instead. Maybe we can enhance globalAttributes to do closest()?

**Review by driskull — APPROVED** [review]
LGTM! Maybe just fix the span elements in the CSS?

File: src/components/slider/resources.ts [hunk 18]
```diff
@@ -1,6 +1,6 @@
 export const CSS = {
-  labelMinValue: "span.handle__label--minValue",
-  labelValue: "span.handle__label--value",
-  tickMin: "span.tick__label--min",
-  tickMax: "span.tick__label--max"
+  labelMinValue: ".handle__label--minValue",
```
Comment (driskull) [inline src/components/slider/resources.ts:2]: @Elijbet these shouldn't have the dot either. `.`

File: src/components/slider/slider.tsx (hunk 17 above)
Comment (jcfranco) [reply src/components/slider/slider.tsx:1433]: This is a compromise right now to avoid querying the DOM between renders. We can doc this limitation, create an issue for adding `lang` lookup and wait for user feedback since users would be expect that to work. cc @geospatialem @benelan

File: .gitignore [hunk 19]
```diff
@@ -33,3 +33,4 @@ UserInterfaceState.xcuserstate
 log.txt
 src/components.d.ts
 src/components/icon/assets
+
```
Comment (jcfranco) [inline .gitignore:36]: This and the `package-lock.json` changes are unrelated, please restore.

File: src/components/slider/slider.e2e.ts [hunk 20]
```diff
@@ -721,64 +730,30 @@ describe("calcite-slider", () => {
...
+          return await getDisplayedValuesArray();
+        });
+        expect(withSeparator).toEqual(formattedValuesArr);
+      };
 
-      withSeparator = await page.$eval("calcite-slider", async (): Promise<string[]> => {
-        return await getDisplayedValuesArray();
-      });
-      expect(withSeparator).toEqual(["2500", "500.000,5", "1000", "1.000.000,5"]);
+      Object.entries(formattedValuesPerLanguageObject);
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:753]: Straggler line, please remove.

File: src/components/slider/slider.e2e.ts [hunk 21]
```diff
@@ -721,64 +730,30 @@ describe("calcite-slider", () => {
...
+      for (const lang of Object.keys(formattedValuesPerLanguageObject)) {
```
Comment (jcfranco) [inline src/components/slider/slider.e2e.ts:754]: This could be simplified to:

```ts
for (const lang in formattedValuesPerLanguageObject) { /...
```

File: src/components/slider/resources.ts [hunk 22]
```diff
@@ -0,0 +1,6 @@
+export const CSS = {
+  labelMinValue: "handle__label--minValue",
```
Comment (jcfranco) [inline src/components/slider/resources.ts:2]: Nice! Next step would be to use this object within the slider's rendering.

**Review by jcfranco — CHANGES_REQUESTED** [review]
